scripts/gen_headers.py

```diff
@@ -0,0 +1,222 @@
+              apisigs=apisigs,
+              coreapis=coreapis,
+              extapis=extapis).
+          encode('utf-8', 'replace'))
+
+        print('Generating cl_d3d10.h...')
+        gen = open(args.directory + '/cl_d3d10.h', 'wb')
+        gen.write(
+          cl_ext_h_template.render_unicode(
+              genExtensions={
+                  'cl_khr_d3d10_sharing',
+                  'cl_intel_sharing_format_query_d3d10',
+              },
+              guard="OPENCL_CL_D3D10_H_",
+              includes=getDisableWarningIncludeString("#include <d3d10.h>"),
+              spec=spec,
+              typedefs=typedefs,
+              macros=macros,
+              structs=structs,
+              enums=enums,
+              apisigs=apisigs,
+              coreapis=coreapis,
+              extapis=extapis).
+          encode('utf-8', 'replace'))
+
+        print('Generating cl_d3d11.h...')
+        gen = open(args.directory + '/cl_d3d11.h', 'wb')
+        gen.write(
+          cl_ext_h_template.render_unicode(
+              genExtensions={
+                  'cl_khr_d3d11_sharing',
+                  'cl_intel_sharing_format_query_d3d11'
+              },
+              guard="OPENCL_CL_D3D11_H_",
+              includes=getDisableWarningIncludeString("#include <d3d11.h>"),
+              spec=spec,
+              typedefs=typedefs,
+              macros=macros,
+              structs=structs,
+              enums=enums,
+              apisigs=apisigs,
+              coreapis=coreapis,
+              extapis=extapis).
+          encode('utf-8', 'replace'))
+
+        print('Generating cl_egl.h...')
+        gen = open(args.directory + '/cl_egl.h', 'wb')
+        gen.write(
+          cl_ext_h_template.render_unicode(
+              genExtensions={'cl_khr_egl_event', 'cl_khr_egl_image'},
+              guard="OPENCL_CL_EGL_H_",
+              spec=spec,
+              typedefs=typedefs,
+              macros=macros,
+              structs=structs,
+              enums=enums,
+              apisigs=apisigs,
+              coreapis=coreapis,
+              extapis=extapis).
+          encode('utf-8', 'replace'))
+
+        print('Generating cl_gl.h...')
+        gen = open(args.directory + '/cl_gl.h', 'wb')
+        gen.write(
+          cl_ext_h_template.render_unicode(
+              genExtensions={
+                  'cl_khr_gl_depth_images',
+                  'cl_khr_gl_event',
+                  'cl_khr_gl_msaa_sharing',
+                  'cl_khr_gl_sharing',
+                  'cl_intel_sharing_format_query_gl',
+              },
+              guard="OPENCL_CL_GL_H_",
+              spec=spec,
+              typedefs=typedefs,
+              macros=macros,
+              structs=structs,
+              enums=enums,
+              apisigs=apisigs,
+              coreapis=coreapis,
+              extapis=extapis).
+          encode('utf-8', 'replace'))
+
+        # cl_layer.h is currently generated but unused because the ICD loader
```

Comment:
This is currently unused due to a difference in the naming convention between the header generation script and the current headers, and specifically the way the current headers are used by the ICD loader.

The header generation script currently generates its function pointers using the convention `clMyFunctionName_fn`, for example:

```c
typedef cl_int (CL_API_CALL *
clGetLayerInfo_fn)(
    size_t param_value_size,
    cl_layer_info param_name,
    void* param_value,
    size_t* param_value_size_ret) ;
```

The current headers and the ICD loader are currently using the convention `pfn_clMyFunctionName` though, at least for layers:

https://github.com/KhronosGroup/OpenCL-ICD-Loader/blob/master/loader/icd.c#L203

I think our options are:

* Don't generate the layers header, at least not now, so code that works with the current headers continues to work.  This is the current option in this PR.
* Switch the generation script to use the `pfn_` convention instead.  This will keep the layers code working, though it may break other code, since the `_fn` convention appears to be most commonly (but not uniformly!) used in the headers.
* Switch the headers and the ICD loader at the same time, and require the ICD loader to be built with the latest headers.
* Try to find a way to pick between the older `pfn_` convention and the newer `_fn` convension so the ICD loader will work with the latest headers and slightly newer headers.  I have some ideas how to do this (the currently cl_layer.h doesn't define `cl_loader_layers` for example, so we could look for that), but admittely these feel like a bit of a hack.

I'm very open to suggestions if we decide on a different option.

Reply:
This had escaped me. Maybe we could specify the convention used in a specific header as a parameter of the template (or maybe later include it in the xml file?)? This would allow the generation to be switched on for all extension headers. Arguably, if we take this route further unifying/refactoring of the convention may prove difficult, so I understand the appeal of 3 above, but this will break packaging workflow once more, and the deprecation of cl_gl_ext burned me a bit.

Reply:
A bit more context on why I am shy on switching the naming convention and never went ahead and do the two pull requests I was talking about above: when we deprecated `cl_gl_ext.h` quite a few users and distribution maintainers workflow broke because they were trying to build older versions of ocl-icd with newer headers, as they wanted to stay on the 2.2 release series and not the update 2.3 version. If we change the headers, the same people will encounter the same problem, twice in a 12 month window is I think too much...

Reply:
Thinking out loud: we could also modify the generation scripts to optionally generate the `pfn_` typedefs in addition to the `_fn` typedefs, either just for the layers (recommended) or globally.  This way we still have uniformity but we retain the `pfn_` typedefs for backwards compatibility.  We could even add something like a `CL_NO_PFN_TYPEDEFS` define to compile these out if we really wanted to, though I don't know that it's worth the bother.

> the deprecation of cl_gl_ext.h burned me a bit.

Yes, I am worried about this.  I'm afraid despite our best intentions this is going to break somebody's workflow.  I've already found and fixed a couple of issues in CTS tests that appeared when I switched to the generated headers ☹️.

Reply:
> Thinking out loud: we could also modify the generation scripts to optionally generate the `pfn_` typedefs in addition to the `_fn` typedefs, either just for the layers (recommended) or globally. This way we still have uniformity but we retain the `pfn_` typedefs for backwards compatibility. We could even add something like a `CL_NO_PFN_TYPEDEFS` define to compile these out if we really wanted to, though I don't know that it's worth the bother.

Sold (generating specifically for `cl_layer.h`).